File: testing/workflows/scripts/apply_translations.mjs
#!/usr/bin/env node
/**
 * Apply translations from JSON back into a DOCX file.
 * Uses docx library to create a new document with translated content.
 * 
 * Note: mammoth is read-only, so we create a simple new DOCX with translations.
 * For full formatting preservation, the original Python approach with python-docx is better.
 */
import { Document, Packer, Paragraph, TextRun, HeadingLevel } from 'docx';
import fs from 'fs/promises';

/**
 * Apply translations and create new DOCX
 */
async function applyTranslations(translationsPath, outputPath) {
  const translationsJson = await fs.readFile(translationsPath, 'utf-8');
  const translations = JSON.parse(translationsJson);
  
  // Build document sections from translations
  const children = [];
  
  for (const item of translations) {
    const text = item.target || item.source;
    const style = item.style || 'body';
    
    if (style === 'heading1') {
      children.push(new Paragraph({
        text: text,
        heading: HeadingLevel.HEADING_1,
      }));
    } else if (style === 'heading2') {
      children.push(new Paragraph({
        text: text,
        heading: HeadingLevel.HEADING_2,
      }));
    } else if (style === 'bold') {
      children.push(new Paragraph({
        children: [new TextRun({ text: text, bold: true })],
      }));
    } else {
      children.push(new Paragraph({ text: text }));
    }
  }
  
  const doc = new Document({
    sections: [{
      properties: {},
      children: children,
    }],
  });
  
  const buffer = await Packer.toBuffer(doc);
  await fs.writeFile(outputPath, buffer);
  
  console.log(`Applied ${translations.length} translations -> ${outputPath}`);
}

async function main() {
  const args = process.argv.slice(2);
  
  if (args.length < 2) {
    console.log('Usage: node apply_translations.mjs <translations.json> <output.docx>');
    process.exit(1);
  }
  
  const [translationsPath, outputPath] = args;
  
  try {
    await applyTranslations(translationsPath, outputPath);
  } catch (error) {
    console.error(`Error: ${error.message}`);
    process.exit(1);
  }
}

main();

File: testing/workflows/scripts/apply_translations.py
#!/usr/bin/env python3
"""Apply translations from JSON back into a DOCX file."""
import json
import re
import sys
from pathlib import Path
from docx import Document

BODY_RE = re.compile(r"^body:p(\d+):r(\d+)$")
HEADER_RE = re.compile(r"^header:s(\d+):p(\d+):r(\d+)$")
FOOTER_RE = re.compile(r"^footer:s(\d+):p(\d+):r(\d+)$")

def apply_translations(docx_path: str, translations_path: str, output_path: str):
    doc = Document(docx_path)
    translations = json.loads(Path(translations_path).read_text(encoding="utf-8"))
    
    applied = 0
    for item in translations:
        unit_id = item["id"]
        target = item.get("target", item.get("source", ""))
        
        run = None
        m = BODY_RE.match(unit_id)
        if m:
            p_i, r_i = map(int, m.groups())
            try:
                run = doc.paragraphs[p_i].runs[r_i]
            except IndexError:
                continue
        
        m = HEADER_RE.match(unit_id)
        if m:
            s_i, p_i, r_i = map(int, m.groups())
            try:
                run = doc.sections[s_i].header.paragraphs[p_i].runs[r_i]
            except IndexError:
                continue
        
        m = FOOTER_RE.match(unit_id)
        if m:
            s_i, p_i, r_i = map(int, m.groups())
            try:
                run = doc.sections[s_i].footer.paragraphs[p_i].runs[r_i]
            except IndexError:
                continue
        
        if run:
            run.text = target
            applied += 1
    
    doc.save(output_path)
    print(f"Applied {applied} translations -> {output_path}")

if __name__ == "__main__":
    if len(sys.argv) < 4:
        print("Usage: python apply_translations.py <input.docx> <translations.json> <output.docx>")
        sys.exit(1)
    
    apply_translations(sys.argv[1], sys.argv[2], sys.argv[3])
